src/main/scala/firrtl/passes/Checks.scala

```diff
@@ -61,6 +61,60 @@ object CheckHighForm extends Pass with LazyLogging {
   class BadPrintfException(x: Char) extends PassException(s"${sinfo}: [module ${mname}] Bad printf format: " + "\"%" + x + "\"")
   class BadPrintfTrailingException extends PassException(s"${sinfo}: [module ${mname}] Bad printf format: trailing " + "\"%\"")
   class BadPrintfIncorrectNumException extends PassException(s"${sinfo}: [module ${mname}] Bad printf format: incorrect number of arguments")
+  class InstanceLoop(loop: String) extends PassException(s"${sinfo}: [module ${mname}] Has instance loop $loop")
+
+  /**
+    * Maintains a one to many graph of each modules instantiated child module.
+    * This graph can be searched for a path from a child module back to one of
+    * it's parents.  If one is found a recursive loop has happened
+    * The graph is a map between the name of a node to set of names of that nodes children
+    */
+  class ModuleGraph {
+    val nodes = new HashMap[String, HashSet[String]]
```

Comment:
It is preferable to not import mutable datastructures by name, instead importing scala.collection.mutable and then instantiate with `new mutable.HashMap[...]`

We have not followed this guideline in the past but I strongly believe we should enforce it in the future.